Deprecate `BoxDecoration.shape`.

bernaferrari opened the pull request:
Issue: https://github.com/flutter/flutter/issues/108051
Discussion: https://github.com/flutter/flutter/pull/103833

Might be my first truly breaking change 🎉.

Need help: I don't know how to get tests working. cc @Hixie 

Examples of errors:

```
00:54 +308: flutter/packages/flutter/../flutter/test/material/chip_test.dart: selected chip and avatar draw darkened layer within avatar circle                       
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following TestFailure was thrown running a test:
Expected: Object or closure painting:
          'a path with Color(0x60191919), that contains [Offset(10.0, 10.0)] and does not contain
[Offset(4.0, 4.0)]'
  Actual: _RenderChip:<_RenderChip#c0abb relayoutBoundary=up10>
   Which: did not match the pattern.
          It called drawPath with a paint whose color, Color(0xff1976d2), was not exactly the
expected color (Color(0x60191919)).
          The stack of the offending call was:
            #0      TestRecordingCanvas.noSuchMethod
(file:///Users/bernardoferrari/Downloads/flutter/packages/flutter/test/rendering/recording_canvas.dart:87:70)
            #1      TestRecordingCanvas.drawPath (dart:ui/painting.dart:4921:8)
```

```
flutter/packages/flutter/../flutter/test/material/switch_test.dart: Switch has default colors when enabled                                         
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following TestFailure was thrown running a test:
Expected: Object or closure painting:
          'a rounded rectangle with Color(0x52000000), RRect: RRect.fromLTRBR(13.0, 17.0, 46.0,
31.0, 7.0)'
          'a circle with Color(0x33000000)'
          'a circle with Color(0x24000000)'
          'a circle with Color(0x1f000000)'
          'a circle with Color(0xfffafafa)'
  Actual: _RenderInkFeatures:<_RenderInkFeatures#61c17>
   Which: did not match the pattern.
          It called 5 other methods on the canvas, the first of which was drawPath(Instance of
'Path', Paint(Color(0x33000000))), but did not call drawCircle() at the time where a circle with
Color(0x33000000) was expected.
```

bernaferrari commented on packages/flutter/test/widgets/box_decoration_test.dart:
```diff
@@ -227,12 +227,8 @@ Future<void> main() async {
-                color: const Color(0xFFFFFFFF),
-              ),
-              shape: BoxShape.circle,
+            decoration: const ShapeDecoration(
```
oh, lol. Fixed.

bernaferrari commented:
The biggest question is how to go from `_RenderChip:<_RenderChip#c0abb relayoutBoundary=up10> ` to `drawCircle` or something like that. If only there were a universal solution for tests. There are not many broken tests (about 20, half in switch), I just don't know how to 'improve them'.

Hixie commented on packages/flutter/lib/src/painting/box_decoration.dart:
```diff
@@ -206,6 +210,10 @@ class BoxDecoration extends Decoration {
   ///
   /// {@macro flutter.painting.BoxDecoration.clip}
+  @Deprecated(
+    'For BoxShape.circle, migrate to ShapeDecoration + CircleBorder(). '
```
For better clarity let's say something `replace BoxDecoration(shape: BoxShape.circle) with ShapeDecoration(shape: const CircleBorder()` or something. Actually we might even be able to automigrate this, cc @Piinks.

(Before we do an automigration we really should see what the performance impact of this change is.)

bernaferrari replied:
Is it possible to verify the performance impact **now**? Ignoring some tests?

Hixie replied:
you can run some of the benchmarks locally before and after, sure. see the `dev/benchmarks` directory.

Hixie commented:
To fix the tests, you'll want to find the expectations that say `paints..circle()..rect()` or whatever, and adjust them to match what we paint now.

Hixie commented:
cc @goderbauer who owns the framework and may have opinions on whether this is worth doing.

bernaferrari commented:
Luckily there seem to be only a few of these cases, but I still have no idea how to go from a RenderThing to paint. Yes, we should adjust, but how do I "print" what a Render is painting so I can adjust the test? I don't know how to do this.

Comparing (https://github.com/flutter/flutter/pull/108052/commits/f4088d19666e5fa4199b90d2d7e1bd2cc01f61f2) `BoxDecoration` with `ShapeDecoration`, I found that `ShapeDecoration` **ALWAYS** adds `Padding` to `Container` (usually zero). Therefore, I made it null when the decorator is used + padding zero. `Container(padding: zero)` will still render as `Padding(Container)` (I don't know why this is wanted, but I'm not changing this behavior).

goderbauer commented:
> who owns the framework and may have opinions on whether this is worth doing.

Seems like a nice cleanup. If it does not have any performance impact, I am for this change.

Piinks commented on packages/flutter/lib/src/widgets/container.dart:
```diff
@@ -374,6 +374,9 @@ class Container extends StatelessWidget {
     }
     final EdgeInsetsGeometry? decorationPadding = decoration!.padding;
     if (padding == null) {
+      if (decorationPadding == EdgeInsets.zero) {
```
Is this a separate change from deprecating BoxDecoration.shape?

bernaferrari replied:
It can be separate. I was trying to get the tests working and found ShapeDecoration adds padding ALWAYS, while BoxDecoration doesn't. This makes ShapeDecoration make one less widget and be on par with BoxDecoration.

There is zero difference, just one less widget on the tree.

bernaferrari replied:
This broke a really random test :( reverted.

Piinks commented on dev/benchmarks/macrobenchmarks/lib/src/web/bench_wrapbox_scroll.dart:
```diff
@@ -97,9 +97,9 @@ class ProductPreview extends StatelessWidget {
             margin: const EdgeInsets.all(23),
             padding: const EdgeInsets.all(18),
-            decoration: const BoxDecoration(
+            decoration: const ShapeDecoration(
```
Dart fix should be able to make this change. There is a fairly new `replacedBy` rule that can replace one class type with another.
So I think in this case, the rule would want to make the change only when `BoxDecoration.shape` is set, have it `replacedBy` `ShapeDecoration`, and then change `shape` from `Box.circle` to `CircleBorder`.

You'll want to account for `BoxShape.rectangular` as well, since some folks may have it explicitly set in their code.

We have a lot of examples of these fix rules in  https://github.com/flutter/flutter/blob/master/packages/flutter/lib/fix_data.yaml, which is where these rules would be added as well.

There is also a wiki page that talks about dart fixes in greater detail: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

Piinks commented on packages/flutter/lib/src/painting/box_decoration.dart:
```diff
@@ -92,6 +92,10 @@ class BoxDecoration extends Decoration {
     this.gradient,
     this.backgroundBlendMode,
+    @Deprecated(
+      'Replace BoxDecoration(shape: BoxShape.circle) with ShapeDecoration(shape: const CircleBorder() '
```
This should probably be more generic, some users may not be using circle.

bernaferrari replied:
BoxShape.rectangle doesn't do anything. I guess, "If BoxShape.rectangle, don't do anything. If BoxShape.circle, replace with ShapeDecoration"?

Hixie replied:
yeah the default is rectangle, and the only other option is circle.

Piinks commented on packages/flutter/lib/src/painting/box_decoration.dart:
```diff
@@ -92,6 +92,10 @@ class BoxDecoration extends Decoration {
     this.backgroundBlendMode,
+    @Deprecated(
+      'Replace BoxDecoration(shape: BoxShape.circle) with ShapeDecoration(shape: const CircleBorder() '
+      'This feature was deprecated after [???]',
```
The deprecation policy has guidance on how to assign this version number: https://github.com/flutter/flutter/wiki/Tree-hygiene#deprecations

Hixie commented:
> Luckily there seem to be only a few of these cases, but I still have no idea how to go from a RenderThing to paint.

Ignore the `Actual:` line, it's just that we can't put the right thing there. It's the last bit that's relevant. It should tell you what is being rendered, and you can change the `paints...` expectations to call the right methods based on what it lists. So far example,  when it says:

```
Expected: Object or closure painting:
          'a rounded rectangle with Color(0x52000000), RRect: RRect.fromLTRBR(13.0, 17.0, 46.0,
31.0, 7.0)'
          'a circle with Color(0x33000000)'
          'a circle with Color(0x24000000)'
          'a circle with Color(0x1f000000)'
          'a circle with Color(0xfffafafa)'
  Actual: _RenderInkFeatures:<_RenderInkFeatures#61c17>
   Which: did not match the pattern.
          It called 5 other methods on the canvas, the first of which was drawPath(Instance of
'Path', Paint(Color(0x33000000))), but did not call drawCircle() at the time where a circle with
Color(0x33000000) was expected.
```

...you want to find where we `expect(..., paints..drawCircle())` and replace it with `expect(..., paints..drawPath())` (because `Which:` comment says it calls drawPath). Does that make sense? The more arguments you can add to these calls the better, too, since it'll test more. Look around for other tests that use `paints` to see what I mean.

bernaferrari commented:
Thanks. I'm doing this right now, already got many tests passing. Running flutter test locally to see what else is missing.

Is there a way to understand what this means? I never understood the "???".
<img width="587" alt="image" src="https://user-images.githubusercontent.com/351125/180317579-5f6e2311-67ad-43be-aa43-51c5de889260.png">

Hixie commented:
>
> Is there a way to understand what this means? I never understood the "???".
> [image: image]
> <https://user-images.githubusercontent.com/351125/180317579-5f6e2311-67ad-43be-aa43-51c5de889260.png>
>
The weird line of gibberish is because your console is not set to UTF-8.

bernaferrari commented:
Tests now pass 😀

Next up: get some benchmark.

bernaferrari commented:
## Benchmarks

We have benchmarks!!!! I made a Switch (the widget that's by far the most affected by this change) many times, applied an animation, and executed the benchmark: 10x, 100x, 1000x, many times.

@Hixie is right. ShapeDecoration can be 20% **slower** than BoxDecoration. But I'm not giving up. I compared ShapeDecoration with BoxDecoration and found the biggest difference. It is not preCache. It is path! DrawPath is slow. DrawCircle is fast. Optimizing `_paintInterior` gave a bit of performance, but it was still much slower. `paintShadows` gave the boost we needed. I'm not killing pre-cache, I'm just changing it to pre-cache the bounds instead of the path (when possible, else do everything else as always).

While optimizing Circle, I optimized Oval and while optimizing Oval I also optimized RoundedRectangle. You know what, I'm changing everything anyway.

First optimized benchmark with Switch (now uses `CircleBorder`):
<img width="1242" alt="image" src="https://user-images.githubusercontent.com/351125/180622951-6b65d8fe-bfea-4a1f-a676-896ef102d602.png">

Then `RoundedRectangleBorder`:
![optimization](https://user-images.githubusercontent.com/351125/180625209-a5038197-4db3-4438-9aeb-c6608bbd1dce.png)

## Implementation details:
I added a function and a getter to ShapeBorder (Hans is going to kill me, twice): `hasOptimizedPaintBox` and `paintBox`. I'm too mentally exhausted to think in better names.
`paintBox` calls `canvas.drawRect`, `canvas.drawCircle`, etc. It "replaces" `getOuterPath` for 80% of the borders. They don't need it anymore (but it is still there, of course).

This is how it works:
- ShapeDecoration wants to draw a circle, paintBox is called.
- ShapeDecoration wants to draw a path, same cache as always.
To control this flow, there is `hasOptimizedPaintBox`. When true, ShapeBorder doesn't cache the Path and calls paintBox (which, if it has drawPath, might be slower, I didn't test, but will work fine). I wish `paintBox` were an optional function that we could check if it exists before caching or not. I don't think Dart allows this, so the second variable is doing this.

As a side effect from my last change, I'm reverting the tests and fixing others.

Instead of changing `circle` to `path`, now we are going from `path` to `rrect`! It seems a great change.

<details>
  <summary>Code I used for benchmark here!</summary>

You can change that function to anything:
```dart
MaterialApp(
      home: Scaffold(
        body: Wrap(children: [
          for (int i = 0; i < 100; i++)
            SizedBox(
                width: 100,
                height: 100,
                child: Center(
                  child: _RotatingWidget(
                    child: Container(
                      width: 50,
                      height: 50,
                      decoration: BoxDecoration(
                        color: Colors.red,
                        // shape: RoundedRectangleBorder(
                        borderRadius: BorderRadius.circular(25),
                      // )
                      ),

                    ),
                  ),
                )),
        ]),
      ),
    );
```
 

```dart
// Copyright 2014 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
import 'package:flutter/material.dart';

import 'recorder.dart';

/// Measures expense of constructing Image widgets.
class BenchBuildCircle extends WidgetRecorder {
  BenchBuildCircle() : super(name: benchmarkName);

  static const String benchmarkName = 'build_circle';

  @override
  Widget createWidget() {
    return MaterialApp(
      home: Scaffold(
        body: Wrap(children: [
          for (int i = 0; i < 1000; i++)
            _RotatingWidget(child: Switch(value: true, onChanged: (value) {})),
        ]),
      ),
    );
  }
}

class _RotatingWidget extends StatefulWidget {
  const _RotatingWidget({required this.child});

  final Widget child;

  @override
  _RotatingWidgetState createState() => _RotatingWidgetState();
}

class _RotatingWidgetState extends State<_RotatingWidget>
    with SingleTickerProviderStateMixin {
  late AnimationController controller;

  @override
  void initState() {
    super.initState();
    controller = AnimationController(
      duration: const Duration(milliseconds: 200),
      vsync: this,
    )..repeat();
  }

  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: controller,
      builder: (BuildContext context, Widget? child) {
        return Transform(
          transform: Matrix4.identity()..rotateZ(2 * 3.14 * controller.value),
          child: widget.child,
        );
      },
    );
  }
}
```

</details>

bernaferrari commented:
I added a default (empty) implementation for `paintBox` function so it doesn't break existing code (and probably most external usage will be limited, no one will re-implement RoundedRectangleBorder).

That's it, for now.

Hixie commented:
This sounds fantastic and I'll definitely look closer.

Thinking out loud, one option might be to have paintBox (or whatever we call it) return a boolean, default implementation returns false, and returning true means you did the work and don't need to use getOuterPath et al. That way you just always call it and don't need to have the decoupled boolean getter.

Hixie commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -533,6 +533,67 @@ abstract class ShapeBorder {
+  ///
+  /// This function provides a default empty implementation as it is only executed with
+  /// [hasOptimizedPaintBox] and so it doesn't break existing code.
+  void paintBox(Canvas canvas, Rect rect, Paint paint, {TextDirection? textDirection}) { }
```
maybe drawOuterPath?

bernaferrari replied:
Let's try.

Hixie commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -533,6 +533,67 @@ abstract class ShapeBorder {
+  ///
+  /// As an example, see how [Path.addRect] became [Canvas.drawRect]:
+  ///
+  /// @override
```
putting this in \`\`\`dart will give syntax highlighting in the API docs and (once my PR lands) will actually do code analysis automatically

bernaferrari replied:
Done, thanks.

bernaferrari commented:
> Thinking out loud, one option might be to have paintBox (or whatever we call it) return a boolean

We need to know in preCache what should be calculated and what shouldn't. So we use the boolean there, and when painting comes we use the boolean again. For shadows it inside a loop, so that wouldn't work :(.

Hixie commented:
Hm yeah that's unfortunate.

bernaferrari commented:
Since we are optimizing things, I also committed  https://github.com/flutter/flutter/pull/108052/commits/e91cb3862ada462cc9cd14e823dae55ca934af2d which calls `drawRect` when possible (when `borderRadius == BorderRadius.zero`). BoxDecoration only checks if `borderRadius == null`). It breaks a single test (chip) that we are refactoring anyway. I've wanted this change for a while, but was slightly breaking.

Hixie commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -533,6 +533,84 @@ abstract class ShapeBorder {
   ///  * [Path.contains], which can tell if an [Offset] is within a [Path].
   Path getInnerPath(Rect rect, { TextDirection? textDirection });
 
+  /// Paint a canvas with appropriate shape.
```
`with the appropriate shape` maybe?

Hixie commented on packages/flutter/lib/src/painting/borders.dart:
````diff
@@ -533,6 +533,84 @@ abstract class ShapeBorder {
   ///  * [Path.contains], which can tell if an [Offset] is within a [Path].
   Path getInnerPath(Rect rect, { TextDirection? textDirection });
 
+  /// Paint a canvas with appropriate shape.
+  ///
+  /// This works as an optimization for [getOuterPath] + [Canvas.drawPath],
+  /// which can be slow. Drawing can be faster by providing a [paintInterior]
+  /// with a specific draw method (instead of a generic [Canvas.drawPath]).
+  /// Examples of specific methods are [Canvas.drawRect] and [Canvas.drawCircle].
+  ///
+  /// A standard paintBox implementation is `canvas.draw` followed by
+  /// what is inside [getOuterPath].
+  ///
+  /// As an example, see how [Path.addRect] became [Canvas.drawRect]:
+  ///
+  /// ```dart
+  /// @override
+  /// Path getOuterPath(Rect rect, { TextDirection? textDirection }) {
+  ///  return Path()
+  ///    ..addRect(rect);
+  /// }
+  ///
+  /// @override
+  /// void paintInterior(Canvas canvas, Rect rect, Paint paint, {TextDirection? textDirection}) {
+  ///  canvas.drawRect(rect, paint);
+  /// }
+  /// ```
+  ///
+  /// When a shape can only be drawn using path, [preferPaintInterior] is recommended
+  /// to be false, so [ShapeBorder] caches the path.
+  ///
+  /// This function provides a default empty implementation as it is only executed
+  /// when [preferPaintInterior] is true.
````
I had a go at rewriting this to use the usual Flutter voice for documentation:

```suggestion
  /// Paint a canvas with the appropriate shape.
  ///
  /// On [ShapeBorder] subclasses whose [preferPaintInterior] method returns true,
  /// this should be faster than using [Canvas.drawPath] with the path provided by
  /// [getOuterPath]. (If [preferPaintInterior] returns false, then this method asserts
  /// in debug mode and does nothing in release mode.)
  ///
  /// Subclasses are expected to implement this method when the [Canvas] API has
  /// a dedicated method to draw the relevant shape. For example, [CircleBorder] uses
  /// this to call [Canvas.drawCircle], and [RoundedRectangleBorder] uses this to
  /// call [Canvas.drawRRect].
  ///
  /// Subclasses that implement this must ensure that calling [paintInterior] is semantically
  /// equivalent to (i.e. renders the same pixels as) calling [Canvas.drawPath] with the
  /// same [Paint] and the [Path] returned from [getOuterPath]), and must also override
  /// [preferPaintInterior] to return true.
  ///
  /// For example, a shape that draws a rectangle might implement [getOutputPath],
  /// [paintInterior], and [preferPaintInterior] as follows:
  ///
  /// ```dart
  /// class RectangleBorder extends ShapeBorder {
  ///   ...
  ///
  ///   @override
  ///   Path getOuterPath(Rect rect, { TextDirection? textDirection }) {
  ///    return Path()
  ///      ..addRect(rect);
  ///   }
  ///
  ///   @override
  ///   void paintInterior(Canvas canvas, Rect rect, Paint paint, {TextDirection? textDirection}) {
  ///    canvas.drawRect(rect, paint);
  ///   }
  ///
  ///   @override
￼  ///   bool get preferPaintInterior => true;
  /// }
  /// ```
  ///
  /// When a shape can only be drawn using path, [preferPaintInterior] should return false.
  /// In that case, classes such as [ShapeDecoration] will cache the path from [getOuterPath]
  /// and call [Canvas.drawPath] directly.
```

Hixie replied:
(i didn't check that the wrapping is correct)

Hixie commented on packages/flutter/lib/src/painting/borders.dart:
```diff
@@ -533,6 +533,84 @@ abstract class ShapeBorder {
+  ///
+  /// This function provides a default empty implementation as it is only executed
+  /// when [preferPaintInterior] is true.
+  void paintInterior(Canvas canvas, Rect rect, Paint paint, {TextDirection? textDirection}) { }
```
I suggest the following default implementation:

```dart
  assert(!preferPaintInterior, '$runtimeType has preferPaintInterior returning true but does not implement paintInterior.');
  assert(false, '$runtimeType.preferPaintInterior returns false, so it is an error to call its paintInterior method.');
```

Hixie commented on packages/flutter/lib/src/painting/borders.dart:
````diff
@@ -533,6 +533,84 @@ abstract class ShapeBorder {
+  /// so [ShapeBorder] caches it. In order to cache, it needs to know
+  /// if [paintInterior] calls [Canvas.drawPath] (the slower method) or
+  /// the faster methods.
+  ///
+  /// Caching a circle will lead to reduced performance,
+  /// so [preferPaintInterior] allows the choice.
+  /// When true, ShapeBorder will reduce caching and call [paintInterior].
+  /// When false, ShapeBorder will cache the path first.
+  ///
+  /// GOOD:
+  /// ```dart
+  /// @override
+  /// void paintInterior(Canvas canvas, Rect rect, Paint paint, { TextDirection? textDirection }) {
+  ///   canvas.drawRect(rect, paint);
+  /// }
+  ///
+  /// @override
+  /// bool get preferPaintInterior => true;
+  /// ```
+  ///
+  /// GOOD:
+  /// ```dart
+  /// @override
+  /// void paintInterior(Canvas canvas, Rect rect, Paint paint, { TextDirection? textDirection }) {
+  ///   canvas.drawPath(_getPath(rect), paint);
+  /// }
+  /// @override
+  /// bool get preferPaintInterior => false;
+  /// ```
+  ///
+  /// BAD:
+  /// ```dart
+  /// @override
+  /// void paintInterior(Canvas canvas, Rect rect, Paint paint, { TextDirection? textDirection }) {
+  ///   canvas.drawPath(_getPath(rect), paint);
+  /// }
+  /// @override
+  /// bool get preferPaintInterior => true;
+  /// ```
+  bool get preferPaintInterior => false;
````
Tried to rewrite this to be more succint:

```suggestion
  /// Reports whether [paintInterior] is implemented.
  ///
  /// Classes such as [ShapeDecoration] prefer to use [paintInterior] if this
  /// getter returns true. This is intended to enable faster painting; instead of
  /// computing a shape using [getOuterPath] and then drawing it using
  /// [Canvas.drawPath], the path can be drawn directly to the [Canvas] using
  /// dedicated methods such as [Canvas.drawRect] or [Canvas.drawCircle].
  ///
  /// By default, this getter returns false.
  ///
  /// Subclasses that implement [paintInterior] should override this to return true.
  /// Subclasses should only override [paintInterior] if doing so enables faster
  /// rendering than is possible with [Canvas.drawPath] (so, in particular, 
  /// subclasses should not call [Canvas.drawPath] in [paintInterior]).
  ///
  /// See also:
  ///
  ///  * [paintInterior], whose API documentation has an example implementation.
  bool get preferPaintInterior => false;
```

Hixie commented:
Some suggestions for the docs but otherwise this looks fantastic.

Any idea why CircleBorder is still a tiny bit slower than BoxDecoration with BoxShape.circle?

bernaferrari commented:
> Any idea why CircleBorder is still a tiny bit slower than BoxDecoration with BoxShape.circle?

Oh, you wanted it a tiny bit faster? Consider it done.
My circle tests are based on Switch which uses shadows. **Apparently** if you replace `shadows.map(shadows.toPaint())` + `shadows.map(shadows.toRect())` with `shadows.map(MapEntry(shadows.toPaint(), shadows.toRect())` it gets faster because it is one less loop (where are tuples in dart?). I can also make a list of lists; which do you prefer? [].key/[].value or [][0], [][1]?

I don't trust my benchmark, I made muuuultiple runs, the values are always changing, and I don't have a controlled environment, but it seems avoiding that loop really improves my test (300x runs with 25 switches rotating).

I also tested calling `shadow.toSide()` on draw time, and seems to be also faster, but caching everything ended up being even better.

<img width="919" alt="image" src="https://user-images.githubusercontent.com/351125/181107531-0c2253ad-b939-4e31-b366-f8eb50e1d89c.png">

Hixie commented:
I love it.

bernaferrari commented:
The quick fix is here, but it is not 100%. If you could help us fix the quick-fix, it would be great @Piinks. My colleague said he doesn't know how to go from `BoxDecoration` to `ShapeDecoration`, even when using your `replaceBy`. If you could take a look, should be something small.

bernaferrari committed:
Fix upstream changes.

Improve wording.

Improve deprecation message.

Revert Quick Fixes

This reverts commit 8b79dacf599f4b362d63d54bc333a9c17baa75eb.

Revert "Fix formatting issue due to incorrect index in fix_data.yaml (#6)"

This reverts commit 3dc9713d2aab2f0e7662ae7729efbfffa51bbad0.

Revert "Fix Quick Fixes"

This reverts commit 777675e8b92584e42be5f21795107ee7d1d29457.
Revert Quick Fixes

Revert Quick Fixes

Minor fixes.

Fix formatting issue due to incorrect index in fix_data.yaml (#6)

Fix analyzer.

Fix Quick Fixes

* Add comments linking to the PR

* Add painting.dart in uris list

* Flutter Fix: account for border and backgroundBlendMode and remove borderRadius.

* Tests for flutter fix

* Flutter fix for rectangle shape in BoxDecoration.copyWith.

* Add flutter fix tests for BoxDecoration.copyWith

* Include rendering.dart for BoxDecoration flutter fix.

Deprecate BoxShape.

Co-authored-by: Birju Vachhani <brvachhani@gmail.com>

bernaferrari commented:
I split this PR in two to make it easier for you in the conflicts that might appear interally. Performance optimization + Deprecation. This is the deprecation part (the quick fix working is what's remaining). The other is already ready for review.

Piinks commented:
> Edit 1: What do you think about renaming boxShadow to shadows? This is the biggest difference between ShapeDecoration and BoxDecoration. Maybe we could deprecate that too?

While that may be cleaner or create parity, we try to avoid changing the name of something just for the aesthetic. If it were a change that facilitated new functionality it would make more sense. Breaking changes and deprecations, in spite of the wonderful tools we have to help, affect a lot of folks, so we try to be very considerate about making them.

I am checking out your quick fixes now - thanks for including them!

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,65 @@
+  - title: 'Remove BoxShape.rectangle'
+    date: 2022-07-27
+    element:
+      uris: [ 'widgets.dart', 'material.dart', 'cupertino.dart' ]
```
Isn't BoxDecoration defined in the painting library? I am trying to find where it gets exported from there, but that may have to do with the fix not working here.

Piinks replied:
The uris here should include every library that exports BoxDecoration. then, each `.dart/.dart.expect` file associated with those uris should contain the sample code to validate each.

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,65 @@
+      shape:
+        kind: 'fragment'
+        value: 'arguments[shape]'
+  - title: 'Replace with ShapeDecoration'
```
Nit:
```suggestion

  - title: 'Replace with ShapeDecoration'
```

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,65 @@
 
 version: 1
 transforms:
+  - title: 'Remove BoxShape.rectangle'
```
Before each title can you add a comment linking to this PR? Other fixes should have the same notation for reference.

Piinks commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,65 @@
+    element:
+      uris: [ 'widgets.dart', 'material.dart', 'cupertino.dart' ]
+      inClass: 'BoxDecoration'
+      constructor: ''
```
As this one if for the constructor, there should also be a fix for the `copyWith` method.

Piinks commented on packages/flutter/test_fixes/box_decoration.dart:
```diff
@@ -0,0 +1,9 @@
+import 'package:flutter/material.dart';
+
+void main() {
+  const BoxDecoration(
+    shape: BoxShape.circle,
+    color: Colors.blue,
+    boxShadow: [],
+  );
```
These should go in the associated library files, instead of individual files. The import of the library is the first designator for applying a fix.

This will need more test cases as well for all of the conditions:

```suggestion
  const BoxDecoration()
  const BoxDecoration(
    shape: BoxShape.circle,
    color: Colors.blue,
    boxShadow: [],
  );
  const BoxDecoration(
    shape: BoxShape.rectangle,
    color: Colors.blue,
    boxShadow: [],
  );
  BoxDecoration().copyWith(shape: BoxShape.circle);
  BoxDecoration().copyWith(shape: BoxShape.rectangle);
```

bernaferrari replied:
> These should go in the associated library files, instead of individual files

Could you please explain better what do you mean? Is there a 'quick fix' directory on `painting`?

Piinks replied:
It does not look like there is one yet, so you could rename these painting.dart and painting.dart.expect. But ensure only the painting library is imported at the top of the file. Since BoxDecoration is exported through other libraries like material, you will need to include the same snippets of code in each of those files as well.

BirjuVachhani commented:
The quick fix to replace `BoxDecoration` with `ShapeDecoration` only works (well I was just testing the implementation visually) when we mark `BoxDecoration` as **deprecated**. In other words, If the tool can find some lint on the `BoxDecoration` constructor then it suggests the quick fix, otherwise it ignores it. I think that's why marking BoxDecoration as deprecated triggers the link and the quick fix.

I even tested this with `dart fix --compare-to-golden .` command when BoxDecoration is marked as deprecated. It passes. But without any lints on BoxDecoration, it fails.

Maybe I am missing something? I'm not sure. @Piinks Would you mind taking a look at the quick fix implementation and see if there's any issue? 

Thanks.

Piinks reviewed (commented):
Catching up on my review queue here, @bernaferrari what is the status of this PR? It looks like there are some merge conflicts and failing checks.

bernaferrari commented:
Hey, @Piinks. The merge conflicts are "fine", the issue is I/we have no idea how to get the quick-fix working. I asked my colleague for help, and he asked yours (https://github.com/flutter/flutter/pull/108052#issuecomment-1229804893), but I don't think you saw that yet :( 

If you could review the quick-fix again and help us get that working, it would be awesome and maybe this time we land the PR. We are struggling with it.

Piinks commented on packages/flutter/test_fixes/widgets.dart:
```diff
@@ -184,4 +185,38 @@ void main() {
+    boxShadow: [],
+  );
+  BoxDecoration(
+    color: Color(0xFFFF0000),
```
Only code that contains an error or triggers a lint would be evaluated for a fix. `shape` is what is being deprecated, so this code that does not contain `shape` would not trigger a fix.

bernaferrari replied:
Thanks!!

bernaferrari commented on packages/flutter/test_fixes/widgets.dart.expect:
```diff
@@ -184,4 +185,35 @@ void main() {
+  BoxDecoration(color: Color(0xFFFF0000), boxShadow: []);
+  BoxDecoration(
+    color: Color(0xFFFF0000),
+    boxShadow: [],
```
`BoxDecoration.boxShadow` becomes `ShapeDecoration.shadows`

so the conversion needs to consider that.

Piinks replied:
Is it necessary though? Does this code still work? There is no analysis or lint that would indicate this is broken code.

Piinks replied:
What I think is missing here, is what exactly is failing? Are some of these transformations working, and others not? From what I can see, the fix rules here do not cover all of these cases.

BirjuVachhani replied:
> Is it necessary though? Does this code still work? There is no analysis or lint that would indicate this is broken code.

Right. The thing is, `shape` is deprecated but the way we can migrate that is by replacing the `BoxDecoration` with `ShapeDecoration` when the shape is circle. 

What's failing is, the fix is not triggered for BoxShape.circle because we are using `replacedBy`.
![image](https://user-images.githubusercontent.com/20423471/190325895-f440e496-b491-48a5-b64e-8d79f20974d9.png)

If I remove this `replacedBy` change, then every thing works except `BoxDecoration` is not renamed to `ShapeDecoration`. So this might be a bug in `replacedBy`? Or I might doing something wrong.

bernaferrari replied:
> Is it necessary though

You are right. Fixed!

BirjuVachhani commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,100 @@
+                uris: [ 'painting.dart', 'rendering.dart', 'widgets.dart', 'material.dart', 'cupertino.dart' ]
+                name: 'CircleBorder'
+        - kind: 'removeParameter'
+          name: 'boxShadow'
```
`boxShadow` is named `shadows` in `ShapeDecoration` api so we need to migrate that too when we convert `BoxDecoration` to `ShapeDecoration`.  If `boxShadow` is not in use then it won't be added as `shadow`. 

> This fix rule is only applied if shape == 'BoxShape.circle' && border == '' && backgroundBlendMode == ''.. is box shadow something that should not be here?

This is because 1. we only apply this fix if the shape is `BoxShape.circle` 2. ShapeDecoration doesn't have `border` and `backgroundBlendMode` params so we only apply if they don't exist in the code.

On second thought, `CircleBorder` can have borderSide but the conversion from `border` would be hard and complicated so I don't know what else to do here.

bernaferrari replied:
IMO you can ignore backgroundBlendMode, it's probably rarely being used and I don't think will break any test.

The only allowed Border with circle is Border.all(), so that could become BorderSide with the same parameters.

BirjuVachhani commented on packages/flutter/lib/fix_data.yaml:
```diff
@@ -17,6 +17,100 @@
+      uris: [ 'painting.dart', 'rendering.dart', 'widgets.dart', 'material.dart', 'cupertino.dart' ]
+      inClass: 'BoxDecoration'
+      constructor: ''
+    oneOf:
```
We only apply the fix in specific condition and there's only 1 option to fix it. But AFAIK, I can't use `if` without using `oneOf`. can I?

bernaferrari commented:
Is it possible to do:
```dart
BoxDecoration(
  shape: BoxShape.circle,
  border: Border.all(width: 20),
)
```
become
```dart
ShapeDecoration(
  side: CircleBorder(width: 20),
)
```
?

I could rename border into side, and add `Border.all()` inside CircleBorder, but is it possible to "remove" Border.all() and keep only what's inside there?

Piinks commented:
> I could rename border into side, and add `Border.all()` inside BorderSide, but is it possible to "remove" Border.all() and keep only what's inside there?

I don't think so. That is too specific and complex for the dart fix tool, and really would only account for a small number of cases. We can't account for every value a user could provide, especially since it could be abstracted into variables and method calls. The dart fix tool is mean to generally migrate folks over to the new API.

bernaferrari commented:
Ok, thanks! Could you help us solve this comment? https://github.com/flutter/flutter/pull/108052#discussion_r971557566
I think this is the last thing missing. I can't make the quick-fix appear, even when simplifying it. Replace-by works different than what I would expect.

Piinks commented:
Yeah I just ran into this myself yesterday. I do not think it is working as expected, I filed https://github.com/dart-lang/sdk/issues/50011 with the dart team to follow up on. You likely won't be able to land that fix right now until they resolve it. Did you want to remove it from this PR, or wait to see what the dart team says?

bernaferrari commented:
Thanks! I knew something was wrong lol.
That's the soul of the quick-fix in this PR, so I think we will need to wait until it is working.

Piinks commented:
Following up on migration guide, there is a link to the template and instructions in our breaking change policy: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes :)

bernaferrari commented:
Ok, if I did it correctly, here it goes: https://github.com/flutter/website/pull/7704

Piinks commented:
Thank you for writing a migration guide! Can you check the analysis failures here?

Piinks commented on packages/flutter/lib/src/painting/box_decoration.dart:
```diff
@@ -92,6 +92,10 @@ class BoxDecoration extends Decoration {
     this.boxShadow,
     this.gradient,
     this.backgroundBlendMode,
+    @Deprecated(
```
We very (like very very) recently decided to try to include some context to help users better understand why we are making a change that requires them to migrate their code. Can you update these deprecation notices to have just a small bit of context, like it being more performant to use ShapeDecoration?

bernaferrari replied:
Done.

```
@Deprecated(
      'Replace BoxDecoration(shape: BoxShape.circle) with ShapeDecoration(shape: const CircleBorder()). '
      'ShapeDecoration is more efficient than BoxDecoration due to caching. '
      'ShapeDecoration also allows code to be tree-shaken, unlike BoxDecoration.shape. '
      'This feature was deprecated after 3.4.0-34.1.pre.',
    )
```

Piinks commented:
It looks like this is very nearly ready 🎉 , will you un-mark it as a draft when you are ready for another review? Thanks!

bernaferrari commented:
Oh, sure. The analyzer was failing due to a missing a "v" in the version 😩.

You might want to check why google testing is failing (and possibly prepare the g-fix for the change). The changes in this PR became really succinct.

Piinks commented:
Hey @bernaferrari can you rebase with the latest? It looks like there is a conflict, and the update will fix the Google check here too.

bernaferrari commented:
Here it goes @Piinks!

Piinks reviewed (commented):
I was just reading through the migration guide, it says the enum BoxShape should be deprecated? Can you add that here?

bernaferrari commented:
"`BoxDecoration.shape` should be deprecated" (the other reviewer complained and I rephrased to "BoxDecoration(shape: BoxShape.circle) should be deprecated").

Unfortunately, we can't deprecate BoxShape because InkWell uses it. But after 1 year it will probably move to another file.

Piinks commented:
> But after 1 year it will probably move to another file.

I am not sure I follow, can you explain?

If we aren't deprecating BoxShape, then we should omit that statement from the migration guide. IT looks like the BoxShape documentation needs updating here as well, it references BoxDecoration. :)

bernaferrari commented:
> I am not sure I follow, can you explain?

BoxShape won't be used anymore in BoxDecoration, but will still be used on InkWell, I don't think we can easily change/"fix" what they are doing there, so we will leave it this way. Since it won't be used in BoxDecoration, we can change to another file where it is used, but we can't do this right now.

> It looks like the BoxShape documentation needs updating here as well

Yes, but it is complicated because, technically, it will still reference until everything is removed.

Piinks commented:
> > It looks like the BoxShape documentation needs updating here as well
>
> Yes, but it is complicated because, technically, it will still reference until everything is removed.

I apologize, I am a bit confused. If we are deciding not to deprecate `boxShape`, then why would it be removed? Currently, the docs for `BoxShape` say:

```
The shape to use when rendering a [Border](https://api.flutter.dev/flutter/painting/Border-class.html) or [BoxDecoration](https://api.flutter.dev/flutter/painting/BoxDecoration-class.html).
```

It is ok to update documentation to recommend the new API, we want folks to migrate away from the old API after all.

bernaferrari commented:
> then why would it be removed?

It will be removed from BoxDecoration, but the class can't be removed because in a random place (InkWell) they use it for different reasons.

I'll update the documentation.

Piinks reviewed (approved):
This LGTM, thank you very much for being so thorough.
We'll need a second review here, I'll see if I can find one. :)

Piinks commented:
It looks like there is a merge conflict and some failing checks, can you take a look? Thanks!

bernaferrari commented:
Here it goes again

Piinks commented:
@bernaferrari just checking in here, I see a second reviewer has not come along so I will reach out again on the discord. The Google testing failure here appears to be a flake with our CI, so it may require rebasing again. With the holidays a lot of folks are out so we might not land this until after new year's.

Really appreciate your contribution, sorry it has taken a lot to get this one over the finish line.

a-wallen reviewed (approved):
LGTM

Piinks commented:
It looks like the failing Google tests are showing circle avatars not being drawn in multiple customer projects. I am not sure why yet, will need to investigate further.

bernaferrari commented:
I unfortunately wasn't able to reproduce :(
If you could get a short snipped that's failing, I would appreciate. It is weird because there was any change that was supposed to cause this.

Renzo-Olivares commented:
Hi @bernaferrari! Would definitely like to help out with these google testing failures if you are still working on this. Do you think you can sync up this PR to master to fix the conflicts?

Renzo-Olivares commented:
I think you have to resync first. The old test results are no longer available.

Renzo-Olivares commented:
Hey @bernaferrari just wanted to check back in to let you know I have been looking at the failing tests, but have not been able to create a minimal reproducible example yet. I'll let you know as soon as I figure it out.

bernaferrari commented:
Ok! Thanks. Looking forward to.

bernaferrari commented:
After exactly 8 months, I think I'll close this. Even though I think `shape: BoxShape` is extreeeemly confusing (since there is the `shape: RoundedRectangleBorder`) and I wish we could rename it for something better (material primary -> foreground was a lot less bad), we are not going to remove it (for now). Any ideas for better names? 

Meanwhile I am, against my will, giving BoxDecoration a reason to exist; Therefore I'm adding non-uniform borders there, since we can't have it in ShapeBorder (natively, at least, for now, at least. My package does it: https://github.com/bernaferrari/non_uniform_border).

https://github.com/flutter/flutter/pull/121921

BirjuVachhani commented:
Yikes! this is taking ages. I was looking forward to this. :(

bernaferrari commented:
Ops, wrong button. I'm just thinking on alternative ways to have the issue "out of the way" and "less confusing" for many people.. `BoxDecoration.circle()` could also be a possibility.

But if anyone figures out the internal Google problem, I'm all in.. I hope you can figure out. But so far, nothing.

Hixie commented:
@bernaferrari I just kicked off a rerun of all the tests, if any of them except Google Testing are failing, we should probably fix those first because otherwise the Google Testing shard will definitely fail anyway (by hitting the same problems).

bernaferrari committed:
Fix calendar test.

Fix

Hixie committed:
Merge branch 'master' into No-BoxShape

Hixie committed:
Merge branch 'master' into No-BoxShape

BirjuVachhani commented:
:eyes:

goderbauer commented:
(triage): @bernaferrari looks like one of the customer tests failed. Do you still have plans to look into that and fix those up?

bernaferrari commented:
I could fix the customer tests, but Google testing is failing on avatar image and no one knows why.

goderbauer commented:
@bernaferrari Do you want to update this PR to the latest master so we can do another google testing run to see what the current status is? Or have you decided to no longer pursue this and we should close it for now?

bernaferrari commented:
Sure, I'll. I just need someone brave enough to be able to see what is happening with AvatarImage lol

MobilyteApps commented:
hi

Piinks commented:
Hey @bernaferrari, we came across this PR during triage today. The problem with AvatarImage has eluded all of us, in spite of multiple attempts to diagnose the issue (over almost 2 years!). Regrettably, I don't know that we will be able to land this PR. Do you think we can close this? Or do you think there is another way we can approach this solution?

I truly appreciate all of the work you have put into this. Unfortunately we have not been able to find a way to land it.

bernaferrari commented:
Yeah, let's close it :( I can't do anything with external help and external help can't figure out.
